### ISSUE TITLE
Not compatible with L.5.5

### ISSUE THREAD

**Issue by mirbaagheri** [neylsongularte/eloquent-extra-events#3]
Hello,
artisan return this error:
`Declaration of NeylsonGularte\EloquentExtraEvents\ExtraEventsTrait::belongsToMany($related, $table = NULL, $foreignKey = NULL, $otherKey = NULL, $relation = NULL) should be com
  patible with Illuminate\Database\Eloquent\Model::belongsToMany($related, $table = NULL, $foreignPivotKey = NULL, $relatedPivotKey = NULL, $parentKey = NULL, $relatedKey = NULL,
   $relation = NULL)`

**Comment by GregPeden** [comment]
I have forked this package and revised for 5.5 support. Please note I have also seriously revised the code structure and, most significantly, the event payload structure is different. Have a look at the code base.

https://github.com/SirLamer/eloquent-extra-events/tree/revise-event-payload

You should bind to a specific commit hash because I have and will make changes as necessary to suit my needs for my work.

I thought the package author here had abandoned this project but I might submit PRs on the stuff I have done since I see an update from a few days ago.

**Comment by neylsongularte** [comment]
@SirLamer could you submit a pull request with 5.5 support, maintaining 5.2, 5.3 and 5.4 support?

**Comment by GregPeden** [comment]
Well, no, and that's the problem.

It's probably better to maintain branches which each support the respective versions. Sprinkling version checks throughout the code and then including different solution paths for each case is not exactly efficient. Among other things, it makes debugging and syntax checking a nightmare.

**Comment by neylsongularte** [comment]
I can launch 5.5 support as version 1.0

**Comment by GregPeden** [comment]
I also changed the format of the event payload, do you want to adopt that as well?

```php

    protected function getBaseEventData($ids)
    {
        return [
            'parent' => $this->getParent(),
            'relationship' => $this->getRelationName(),
            'related' => get_class($this->getRelated()),
            'ids' => static::processIds($ids)
        ];
    }
```

And of course some events add to this with the payload from their respective actions.

**Comment by neylsongularte** [comment]
alright for me

**Comment by neylsongularte** [comment]
What is the content of 'relationship'?

**Comment by GregPeden** [comment]
It is the name of the relationship being queried, while 'related' is the model (class name) of the Eloquent model collection being returned.

So for example, a relationship 'members' may return a collection of Users and a relationship 'banned' may also return Users. The context of each relationship collection is important.

**Comment by GregPeden** [comment]
For the 'related' property, I have half a mind to actually return the result from Laravel's morph map, instead of always the class name. Note that when there is no morph defined, the returned value is in fact the fully qualified class name, but then this provides developers some easily flexibility and it also promotes smaller data packages when doing things like event broadcasting. This would also mirror the behavor for polymorphic relationships, which use the morph map for the 'type' column.

**Comment by mirbaagheri** [comment]
I cloned your project @neylsongularte.
That work and compatible with Laravel 5.5
https://github.com/mostafamiri65/eloquent-extra-events

**Comment by GregPeden** [comment]
I prepared a 5.5 branch from my work. As a reminder, I made many changes beyond just 5.5 support.
https://github.com/SirLamer/eloquent-extra-events/tree/5.5

The branch above actually did not include another feature I provided: Event register methods for your Eloquent model which uses the trait.

I would be happy if you want me to submit this as a PR however at this point it is nearly a total re-write, so if I send this as a PR it will pretty much replace your project with this version. So I'll do so if you indicate that you intend to continue to maintain this code base, otherwise I can continue on myself. No pressure, I just want to make sure the best choice for future support is made.
